kola/tests/misc/network: Add dhcp test with veth

RamLavi opened the pull request:
This PR is adding a test that will check DHCP propagation over a bond of two veth pairs.

test network chart attached 
![test-dhcp-netns](https://user-images.githubusercontent.com/48101085/122675497-5b545680-d1e2-11eb-819b-05e0ef503027.png)

RamLavi commented:
I'm getting a docker command not found error, which is weird since docker is used on other tests such as [docker test](https://github.com/coreos/coreos-assembler/blob/main/mantle/kola/tests/docker/docker.go#L229).
```
$ make && ./bin/kola run -b rhcos --qemu-image /tmp/dhcp_test/rhcos-4.7.7-x86_64-qemu.x86_64.qcow2 rhcos.network.bondWithDhcp || loud
./build cmd/*
Building kola
go build: -i flag is deprecated
Building kolet (static)
Building ore
go build: -i flag is deprecated
Building plume
go build: -i flag is deprecated
=== RUN   rhcos.network.bondWithDhcp
--- FAIL: rhcos.network.bondWithDhcp (67.78s)
        cluster.go:141: sudo: docker: command not found
        cluster.go:164: "tmpdir=$(mktemp -d); cd $tmpdir; echo -e \"FROM scratch\\nRUN dnf -y install systemd dnsmasq iproute iputils && dnf clean all && systemctl enable dnsmasq\\nRUN echo -e \"dhcp-range=192.168.0.50,192.168.0.60,12h\\ndhcp-host=52:55:00:d1:56:00,192.168.0.55\" > /etc/dnsmasq.d/dhcp && systemctl restart dnsmasq\" > Dockerfile;\n\t      sudo docker build -t dnsmaq ." failed: output , status Process exited with status 1

```
@jlebon @darkmuggle , do you happen to know why?

RamLavi commented:
> I'm getting a docker command not found error, which is weird since docker is used on other tests such as [docker test](https://github.com/coreos/coreos-assembler/blob/main/mantle/kola/tests/docker/docker.go#L229).
> 
> ```
> $ make && ./bin/kola run -b rhcos --qemu-image /tmp/dhcp_test/rhcos-4.7.7-x86_64-qemu.x86_64.qcow2 rhcos.network.bondWithDhcp || loud
> ./build cmd/*
> Building kola
> go build: -i flag is deprecated
> Building kolet (static)
> Building ore
> go build: -i flag is deprecated
> Building plume
> go build: -i flag is deprecated
> === RUN   rhcos.network.bondWithDhcp
> --- FAIL: rhcos.network.bondWithDhcp (67.78s)
>         cluster.go:141: sudo: docker: command not found
>         cluster.go:164: "tmpdir=$(mktemp -d); cd $tmpdir; echo -e \"FROM scratch\\nRUN dnf -y install systemd dnsmasq iproute iputils && dnf clean all && systemctl enable dnsmasq\\nRUN echo -e \"dhcp-range=192.168.0.50,192.168.0.60,12h\\ndhcp-host=52:55:00:d1:56:00,192.168.0.55\" > /etc/dnsmasq.d/dhcp && systemctl restart dnsmasq\" > Dockerfile;\n\t      sudo docker build -t dnsmaq ." failed: output , status Process exited with status 1
> ```
> 
> @jlebon @darkmuggle , do you happen to know why?

swapped docker for podman, now I'm getting a `stat /etc/resolv.conf: no such file or directory` error when I try to `dnf install`
```
$ make && ./bin/kola run -b rhcos --qemu-image /tmp/dhcp_test/rhcos-4.7.7-x86_64-qemu.x86_64.qcow2 rhcos.network.bondWithDhcp || loud
./build cmd/*
Building kola
go build: -i flag is deprecated
Building kolet (static)
Building ore
go build: -i flag is deprecated
Building plume
go build: -i flag is deprecated
=== RUN   rhcos.network.bondWithDhcp
--- FAIL: rhcos.network.bondWithDhcp (85.06s)
        cluster.go:141: Error: error building at STEP "RUN dnf -y install systemd": stat /etc/resolv.conf: no such file or directory
        cluster.go:164: "tmpdir=$(mktemp -d); cd $tmpdir; echo -e \"FROM scratch\\nRUN dnf -y install systemd\" > Dockerfile;\n\t        sudo podman build -t dnsmaq ." failed: output STEP 1: FROM scratch
STEP 2: RUN dnf -y install systemd, status Process exited with status 125
FAIL, output in _kola_temp/qemu-unpriv-2021-06-09-1938-339956
Error: harness: test suite failed
2021-06-09T16:40:09Z cli: harness: test suite failed
```

RamLavi commented:
the former issue with resolve.conf suggests that `FROM scratch` is too minimalistic for my needs.
However, I seem to not be able to pull the `registry.fedoraproject.org/fedora:33` image: 
```
$ make && ./bin/kola run -b rhcos --qemu-image /tmp/dhcp_test/rhcos-4.7.7-x86_64-qemu.x86_64.qcow2 rhcos.network.bondWithDhcp || loud
./build cmd/*
Building kola
go build: -i flag is deprecated
Building kolet (static)
Building ore
go build: -i flag is deprecated
Building plume
go build: -i flag is deprecated
=== RUN   rhcos.network.bondWithDhcp
--- FAIL: rhcos.network.bondWithDhcp (101.04s)
        cluster.go:141: Error: error creating build container: Error initializing source docker://registry.fedoraproject.org/fedora:33: error pinging docker registry registry.fedoraproject.org: Get "https://registry.fedoraproject.org/v2/": dial tcp: lookup registry.fedoraproject.org on [::1]:53: read udp [::1]:58962->[::1]:53: read: connection refused
        cluster.go:164: "tmpdir=$(mktemp -d); cd $tmpdir; echo -e \"FROM registry.fedoraproject.org/fedora:33\" > Dockerfile;\n\t        sudo podman build -t dnsmaq ." failed: output STEP 1: FROM registry.fedoraproject.org/fedora:33, status Process exited with status 125
FAIL, output in _kola_temp/qemu-unpriv-2021-06-10-1342-366257
Error: harness: test suite failed
```
@jlebon @darkmuggle do you know a way around this? is this a networking issue?

darkmuggle commented:
Pulling your PR, here's what I'm getting locally:
```
=== RUN   rhcos.network.bondWithDhcp
2021-06-10T15:25:58Z platform: Started qemu (18145) with args: [qemu-system-x86_64 -machine accel=kvm -cpu host -m 1024 -smp 1 -object rng-random,filename=/dev/urandom,id=rng0 -device virtio-rng-pci,rng=rng0 -uuid f2a1efae-7945-4ae0-a1e0-92f871525529 -nographic -boot order=c,strict=on -chardev socket,id=chrtpm,path=/var/tmp/mantle-qemu239739576/swtpm-sock -tpmdev emulator,id=tpm0,chardev=chrtpm -device tpm-tis,tpmdev=tpm0 -add-fd fd=3,set=1 -add-fd fd=4,set=2 -device virtio-blk-pci,drive=d1,serial=primary-disk,bootindex=1 -drive if=none,id=d1,file=/dev/fdset/1,auto-read-only=off,cache=unsafe -fw_cfg name=opt/com.coreos/config,file=tmp/kola/qemu-unpriv-2021-06-10-1525-18121/rhcos.network.bondWithDhcp/f2a1efae-7945-4ae0-a1e0-92f871525529/ignition.json -netdev user,id=eth0,hostfwd=tcp:127.0.0.1:43645-:22,hostname=qemu0,restrict=on -device virtio-net-pci,netdev=eth0 -chardev socket,id=qemu-qmp,path=/var/tmp/mantle-qemu239739576/qmp.sock,server,nowait -mon chardev=qemu-qmp,mode=control -device virtio-serial -chardev file,id=virtioserial1,path=/dev/fdset/2,append=on -device virtserialport,chardev=virtioserial1,name=com.coreos.ignition.journal -display none -chardev file,id=log,path=tmp/kola/qemu-unpriv-2021-06-10-1525-18121/rhcos.network.bondWithDhcp/f2a1efae-7945-4ae0-a1e0-92f871525529/console.txt -serial chardev:log]
2021-06-10T15:27:06Z platform: Killing qemu (18145)
--- FAIL: rhcos.network.bondWithDhcp (69.22s)
        network.go:569: failed to reboot the machine: machine "f2a1efae-7945-4ae0-a1e0-92f871525529" failed basic checks: some systemd units failed:
capture-macs.service loaded failed failed Capture MAC address from kargs
FAIL, output in tmp/kola/qemu-unpriv-2021-06-10-1525-18121
Error: harness: test suite failed
2021-06-10T15:27:07Z cli: harness: test suite failed
[coreos-assembler]$ kola run -d rhcos.network.bondWithDhcp 
2021-06-10T15:29:47Z cli: Started logging at level DEBUG
2021-06-10T15:29:47Z cli: Started logging at level DEBUG
```

jlebon commented on mantle/kola/tests/misc/network.go:
```diff
@@ -294,19 +305,196 @@ var (
+
+// make a podman container out of binaries on the host
+func genDHCPServerContainer(c cluster.TestCluster, m platform.Machine) {
+	//cmd := `tmpdir=$(mktemp -d); cd $tmpdir; echo -e "FROM scratch\nRUN dnf -y install systemd dnsmasq iproute iputils && dnf clean all && systemctl enable dnsmasq\nRUN echo -e "dhcp-range=192.168.0.50,192.168.0.60,12h\ndhcp-host=52:55:00:d1:56:00,192.168.0.55" > /etc/dnsmasq.d/dhcp && systemctl restart dnsmasq" > Dockerfile;
```
The `systemctl restart dnsmasq` looks incorrect there. We want to build the image in this step, not run dnsmasq itself. Compare with https://github.com/coreos/fedora-coreos-config/blob/da83a0aa8d9cadad4ee132e7a7af1f9e88d08b59/tests/kola/chrony/dhcp-propagation#L37-L47.

jlebon commented on mantle/kola/tests/misc/network.go:
```diff
@@ -294,19 +305,196 @@ var (
+
+	cmd := `tmpdir=$(mktemp -d); cd $tmpdir; echo -e "FROM registry.fedoraproject.org/fedora:33" > Dockerfile;
+	        sudo podman build -t dnsmaq .`
+	c.MustSSH(m, cmd)
```
I.e. we're missing a `podman run` around here.

jlebon commented on mantle/kola/tests/misc/network.go:
```diff
@@ -59,6 +61,15 @@ func init() {
+		Run:         NetworkBondWithDhcp,
+		ClusterSize: 0,
+		Name:        "rhcos.network.bondWithDhcp",
+		Distros:     []string{"rhcos"},
```
I think you need something like:

```
Tags:             []string{kola.NeedsInternetTag},
```

here. Otherwise by default we restrict internet access. See e.g. the Tang tests in `luks.go` as examples.

RamLavi replied:
Thanks! but I think that after the change i pushed it will no longer be need.

RamLavi commented:
@darkmuggle @jlebon 
In order to reduce complexity, I changed the test a bit to run the dnsmasq inside a network namespace instead of from a container. It basically does the same thing, allowing the dnsmasq to serve the two host-side veth ends, but without using docker/external internet connection.
I also moved the the call of the setup-ovs script to run within the script, so that it can use the same dnsmasq process while setting the `brcnv-iface` up.
I checked the test locally and it looks stable.

I would appreciate your review when available.

jlebon commented on mantle/kola/tests/misc/network.go:
```diff
@@ -300,10 +300,10 @@ func NetworkSecondaryNics(c cluster.TestCluster) {
-func checkOvsBridge(c cluster.TestCluster, setPrimaryMac string) {
+func checkOvsBridge(c cluster.TestCluster, setPrimaryMac, setPrimaryIp string) {
 	var err error
 	machineIdx := 0
```
Not new, but I don't think we need this, right? We only provision one machine so the right one is always going to be 0. Then you can avoid needing a parameter for that in `getBondIfaceIP` (or if you want to keep it generic, passing a `platform.Machine` would be a bit cleaner).

RamLavi replied:
totally agree. Done

jlebon commented on mantle/kola/tests/misc/network.go:
```diff
@@ -56,6 +56,15 @@ func init() {
+	register.RegisterTest(&register.Test{
+		Run:         NetworkBondWithDhcp,
+		ClusterSize: 0,
+		Name:        "rhcos.network.bondWithDhcp",
```
Minor/optional: our naming scheme for tests is usually kebab-casing (see `kola list`).

RamLavi replied:
np. Done

jlebon reviewed (commented):
> I'm not sure why `continuous-integration/jenkins/pr-merge` lane is failing - I didn't find any explanation other than `Error build image: status phase Failed`. am I looking in the wrong place?

Yeah it's kinda confusing. The output from the image build is actually in the step labelled "Internal utility function for OpenShift DSL". There you'll see `gofmt` tripping. Or you can also look at the full job output for the `Build` stage: https://jenkins-coreos-ci.apps.ocp.ci.centos.org/blue/rest/organizations/jenkins/pipelines/github-ci/pipelines/coreos/pipelines/coreos-assembler/branches/PR-2221/runs/3/nodes/52/log/?start=0

jlebon commented:
/ok-to-test

jlebon commented:
Anyway, overall
/approve
!

Thanks for splitting up the commits nicely.

RamLavi commented:
@jlebon thanks for the review, fixes [here](https://github.com/coreos/coreos-assembler/compare/87614e92badcf36baaf17d3296374d01ebb8656e..372ee2c659a1ce303569b41215efe66d452348a3)

phoracek commented on mantle/kola/tests/misc/network.go:
```diff
@@ -219,6 +228,67 @@ var (
+
+            # Assign an IP address to netns-side veth end and bring it up
+            ip netns exec ${network_namespace} ip address add ${netns_side_ip} dev ${veth_netns_side_end_name}
+            ip netns exec ${network_namespace} ip -6 addr flush ${veth_netns_side_end_name}
```
Is this needed?

RamLavi replied:
forgot to remove this. thanks! Done

phoracek commented on mantle/kola/tests/misc/network.go:
```diff
@@ -219,6 +228,67 @@ var (
+          activate_veth_end() {
+            veth_end_name=$1
+            nmcli dev set ${veth_end_name} managed yes
+            ip link set ${veth_end_name} up
```
If you use nmcli above, why not here?

RamLavi replied:
well I didn't find a one liner to replace this line.
```
$ nmcli con s
NAME                UUID                                  TYPE      DEVICE     
Wired connection 1  b7e6f4f5-6893-3ad4-bcd5-663cecfff7a7  ethernet  ens5       
Wired connection 2  81c099c1-ff96-3d2d-aa01-0c9db86c48ac  ethernet  veth1-host 
Wired connection 3  1f3d8a11-9aeb-3165-b3a8-7b879db21972  ethernet  --         
[core@cosa-devsh ~]$ nmcli con up 'Wired connection 3'
Error: Connection activation failed: No suitable device found for this connection (device ens5 not available because profile is not compatible with device (mismatching interface name)).
```

phoracek replied:
I guess you are leveraging the side-effect that when a NIC is managed and brought UP, NM will create a profile for it. IMO it would be cleaner if you just created the profile explicitly and stuck to nmcli commands instead of mixing them with ip link.

RamLavi replied:
I agree. I think the most straight forward way is `sudo nmcli con up id 'Wired connection 3'` but that's giving me an error I don't understand (see above)..
I am not finding the correct command to do that. Can you hint me to the right way to do it?

RamLavi replied:
Hey, I think that the reason I'm getting this error is because network manager does not fully support veths in this version
IIUC, nmcli supports veth only from [network-manager 1.30](https://networkmanager.dev/blog/networkmanager-1-30/)
which AFAIK is used only on rhel 8.4 (our current rhcos image is forked from 8.3)
```
[core@cosa-devsh ~]$ nmcli --version
nmcli tool, version 1.26.0-14.1.rhaos4.7.el8
[core@cosa-devsh ~]$ cat /etc/redhat-release 
Red Hat Enterprise Linux CoreOS release 4.7
[core@cosa-devsh ~]$ cat /etc/os-release 
NAME="Red Hat Enterprise Linux CoreOS"
VERSION="47.83.202105220305-0"
VERSION_ID="4.7"
OPENSHIFT_VERSION="4.7"
RHEL_VERSION="8.3"
PRETTY_NAME="Red Hat Enterprise Linux CoreOS 47.83.202105220305-0 (Ootpa)"
ID="rhcos"
ID_LIKE="rhel fedora"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:redhat:enterprise_linux:8::coreos"
HOME_URL="https://www.redhat.com/"
BUG_REPORT_URL="https://bugzilla.redhat.com/"
REDHAT_BUGZILLA_PRODUCT="OpenShift Container Platform"
REDHAT_BUGZILLA_PRODUCT_VERSION="4.7"
REDHAT_SUPPORT_PRODUCT="OpenShift Container Platform"
REDHAT_SUPPORT_PRODUCT_VERSION="4.7"
OSTREE_VERSION='47.83.202105220305-0'
```

phoracek replied:
Sound good

phoracek commented on mantle/kola/tests/misc/network.go:
```diff
@@ -219,6 +228,67 @@ var (
+          create_veth_pair ${veth2_host_side_end_name} ${veth2_netns_side_end_name} ${secondary_mac} 192.168.0.60 ${network_namespace}
+
+          # Run a dnsmasq service on the network_namespace, to set the host-side veth ends a ip via their MAC addresses
+          echo -e "dhcp-range=192.168.0.50,192.168.0.60,255.255.255.0,12h\ndhcp-host=${primary_mac},${primary_ip}\ndhcp-host=${secondary_mac},${secondary_ip}" > /etc/dnsmasq.d/dhcp
```
The server range overlaps with the server IP, this way we would give away the DHCP server IP. While the server should be within the subnet, I don't think it should be in the range

RamLavi replied:
not sure what you mean here. I don't see that the dnsmasq server got an ip, can you elabrate?:
```
dnsmasq: started, version 2.79 cachesize 150
dnsmasq: compile time options: IPv6 GNU-getopt DBus no-i18n IDN2 DHCP DHCPv6 no-Lua TFTP no-conntrack ipset auth DNSSEC loop-detect inotify
dnsmasq-dhcp: DHCP, IP range 192.168.0.50 -- 192.168.0.60, lease time 12h+ ip link set veth1-host up
dnsmasq: reading /etc/resolv.conf
dnsmasq: using nameserver 10.0.2.3#53
dnsmasq: read /etc/hosts - 2 addresses
dnsmasq-dhcp: DHCPDISCOVER(veth1-netns) 52:55:00:d1:56:00 
dnsmasq-dhcp: DHCPOFFER(veth1-netns) 192.168.0.55 52:55:00:d1:56:00 
dnsmasq-dhcp: DHCPREQUEST(veth1-netns) 192.168.0.55 52:55:00:d1:56:00 
dnsmasq-dhcp: DHCPACK(veth1-netns) 192.168.0.55 52:55:00:d1:56:00 
dnsmasq: reading /etc/resolv.conf
dnsmasq: using nameserver 10.0.2.3#53
dnsmasq: ignoring nameserver 192.168.0.59 - local interface
dnsmasq-dhcp: DHCPDISCOVER(veth2-netns) 52:55:00:d1:56:01 
dnsmasq-dhcp: DHCPOFFER(veth2-netns) 192.168.0.56 52:55:00:d1:56:01 
dnsmasq-dhcp: DHCPREQUEST(veth2-netns) 192.168.0.56 52:55:00:d1:56:01 
dnsmasq-dhcp: DHCPACK(veth2-netns) 192.168.0.56 52:55:00:d1:56:01 
dnsmasq: reading /etc/resolv.conf
dnsmasq: using nameserver 10.0.2.3#53
dnsmasq: ignoring nameserver 192.168.0.59 - local interface
dnsmasq: ignoring nameserver 192.168.0.60 - local interface
```

phoracek replied:
Correct me if I'm wrong:

dhcp-range defines the range of IPs that will be served to clients
Your DHCP server is serving on interfaces with IPs 192.168.0.59 and 192.168.0.60 which happens to be in the range of offered IPs for clients. If you call the server with enough clients, you would assign 192.168.0.59 and 192.168.0.60 to some of the clients and that would lead to duplicated IPs on the network.

RamLavi replied:
oh I understand, you mean the netns-side veth ends. 
I figured we won't have any other dhcp requests, but you're right, better put them outside of the range.
Done

phoracek commented on mantle/kola/tests/misc/network.go:
```diff
@@ -219,6 +228,67 @@ var (
+          activate_veth_end ${veth1_host_side_end_name}
+          activate_veth_end ${veth2_host_side_end_name}
+          # Wait to let the dhcp request to be processed
+          sleep 10
```
Could we use polling for the profile to be activated instead? The sleep could lead to races

RamLavi replied:
sure. Done

phoracek commented on mantle/kola/tests/misc/network.go:
```diff
@@ -293,35 +363,141 @@ var (
+			]
+		}
+	}`,
+		base64.StdEncoding.EncodeToString([]byte(defaultLinkConfig)),
```
Do we need all these scripts in this test?

RamLavi replied:
I was keeping them in order to be as close to the target configuration. If some of the scripts may interfere in the future in some unforeseeable way, we would catch it. 
But if you think it's not relevant in any way then I can remove.

phoracek replied:
We should either test it knowing what is the overlap or not not at all. If some of those scripts take effect and influence the tested script, it's fine

RamLavi replied:
not sure I got you here, could you elaborate?

phoracek replied:
Putting all the scripts on one pile to check that things work fine "just in case" sounds odd to me. I would like us to understand whether there is a possible overlap first and then make an educated decision. This set of tests does not replicate exactly what would happen in production, so it is almost as if we were testing integration between integration tests

RamLavi replied:
ack :)
From what I gather:
- `defaultLinkConfig` script is making sure that the nic are named according to their mac-address
since we are not relying on this in our tests, then I think we can drop it from both tests.
- `dhcpClientConfig` - we can drop it for multiple-nics test since we don't use dhcp there. in the dhcp test we should use it - only to make sure that we are consistently using the same dhcp client.
Done

phoracek commented on mantle/kola/tests/misc/network.go:
```diff
@@ -293,35 +363,141 @@ var (
 	setupMultipleNetworkTest(c, primaryMac, secondaryMac)
 
-	checkOvsBridge(c, primaryMac)
+	checkOvsBridge(c, primaryMac, "")
```
Could you create a new function for this instead? One that would set the primary IP, one that would not care about it. They can use the same helper function inside. Having dedicated function names would make it more clear and readable.

RamLavi replied:
changed the function to `checkOvsBridgeParams`, then created 2 wrapper functions `checkBondReceivedExpectedMAC` `checkBondReceivedExpectedMACAndIP`. 
Is it readable enough?

phoracek commented on mantle/kola/tests/misc/network.go:
```diff
@@ -293,35 +363,141 @@ var (
 
-func getBondIfaceMAC(c cluster.TestCluster, machineIdx int, interfaceName string) (string, error) {
-	m := c.Machines()[machineIdx]
+	if setPrimaryIp != "" {
```
Can the content of this be moved to a separate function, so we don't bloat this one?

RamLavi replied:
moved each to a functions to avoid bloating. Done

RamLavi commented:
@phoracek fixed some of the comments [here](https://github.com/coreos/coreos-assembler/compare/372ee2c659a1ce303569b41215efe66d452348a3..ec0eb036e68dafc181b31f2031dff47978fc1d29) and commented back on the rest.

RamLavi commented:
@phoracek moved the ips of the netns-veth ends outside of the dnsmasq range [here](https://github.com/coreos/coreos-assembler/compare/ec0eb036e68dafc181b31f2031dff47978fc1d29..11f09e2be2c6fe635fbd149293ebfb71deadd383) - thanks for the clarification :) 

regarding the other 2 comments - I'm afraid I still didn't catch your drift. Let's discuss further.

phoracek commented on mantle/kola/tests/misc/network.go:
```diff
@@ -219,6 +229,81 @@ var (
+            veth_end_name=$1
+            for attempt in {1..5}; do
+              cidr=$(nmcli -g IP4.ADDRESS dev show ${veth_end_name})
+              if [[ ! -z ${cidr} ]]; then
```
This will fail if the output is empty. You have to put quotes around the variable

RamLavi replied:
Done

phoracek commented on mantle/kola/tests/misc/network.go:
```diff
@@ -219,6 +229,81 @@ var (
+
+          # Run a dnsmasq service on the network_namespace, to set the host-side veth ends a ip via their MAC addresses
+          echo -e "dhcp-range=192.168.0.50,192.168.0.60,255.255.255.0,12h\ndhcp-host=${primary_mac},${primary_ip}\ndhcp-host=${secondary_mac},${secondary_ip}" > /etc/dnsmasq.d/dhcp
+          ip netns exec ${network_namespace} dnsmasq --log-queries --log-facility=/var/log/dnsmasq.log &
```
How do we recover these logs?

RamLavi replied:
When I was debugging using this, I just added this
```
output := string(c.MustSSH(m, "cat /var/log/dnsmasq.log"))
c.Fatal(output)
```
it makes the test temporarily fail, but it's good enough for debugging imo

phoracek replied:
Do we need to carry dead code there then? If you have to make changes to the code to use it, why would not you do what you describe above and add the `--log-facility`

RamLavi replied:
I agree :) Done

RamLavi commented:
Hey @phoracek latest fixes [here](https://github.com/coreos/coreos-assembler/compare/11f09e2be2c6fe635fbd149293ebfb71deadd383..f5d5b605eaf91e31d492d16c0384e5e8002929b1)

RamLavi commented:
hey @jlebon 
I'm not sure why the pr-merge lane is failing again - do you think it's related to me change?
```
[2021-06-24T13:18:20.432Z] [logs:build/coreos-ci-coreos-assembler-ee1083f5-da76-49a9-ae20-9c73254b2ed8-1] go: google.golang.org/api@v0.5.0: unrecognized import path "google.golang.org/api": reading https://google.golang.org/api?go-get=1: 500 Internal Server Error

[2021-06-24T13:18:20.432Z] [logs:build/coreos-ci-coreos-assembler-ee1083f5-da76-49a9-ae20-9c73254b2ed8-1] make[1]: *** [Makefile:11: build] Error 1

[2021-06-24T13:18:20.432Z] [logs:build/coreos-ci-coreos-assembler-ee1083f5-da76-49a9-ae20-9c73254b2ed8-1] make[1]: Leaving directory '/root/containerbuild/tools'

[2021-06-24T13:18:20.432Z] [logs:build/coreos-ci-coreos-assembler-ee1083f5-da76-49a9-ae20-9c73254b2ed8-1] make: *** [Makefile:74: tools] Error 2

[2021-06-24T13:21:54.642Z] [logs:build/coreos-ci-coreos-assembler-ee1083f5-da76-49a9-ae20-9c73254b2ed8-1] error: build error: error building at STEP "RUN ./build.sh make_and_makeinstall": error while running runtime: exit status 2
```

jlebon commented:
> hey @jlebon
> I'm not sure why the pr-merge lane is failing again - do you think it's related to my change?

Nah, looks like infra flake. You can restart the test, unless you'll push more changes anyway.

RamLavi commented on mantle/kola/tests/misc/network.go:
```diff
@@ -293,35 +373,142 @@ var (
+	checkOvsBridgeParams(c, setPrimaryMac, "")
+}
 
+func checkBondReceivedExpectedMACAndIP(c cluster.TestCluster, setPrimaryMac, setPrimaryIp string) {
```
Done

RamLavi commented on mantle/kola/tests/misc/network.go:
```diff
@@ -293,35 +373,142 @@ var (
+	checkOvsBridgeParams(c, setPrimaryMac, setPrimaryIp)
+}
+
+func checkOvsBridgeParams(c cluster.TestCluster, setPrimaryMac, setPrimaryIp string) {
```
Done

RamLavi commented on mantle/kola/tests/misc/network.go:
```diff
@@ -293,35 +373,142 @@ var (
+	}
+}
+
+func checkBondIfaceMAC(c cluster.TestCluster, m platform.Machine, ovsBridgeName, setPrimaryMac string) {
```
Done

RamLavi commented on mantle/kola/tests/misc/network.go:
```diff
@@ -334,6 +521,29 @@ func getBondIfaceMAC(c cluster.TestCluster, machineIdx int, interfaceName string
 	return macAddress.String(), nil
 }
 
+func checkBondIfaceIP(c cluster.TestCluster, m platform.Machine, ovsBridgeName, setPrimaryIp string) {
```
Done

RamLavi commented:
Hey @phoracek 
refactored the second commit and removed the `checkOvs` function. using now `checkExpectedMAC` & `checkExpectedIP`
changes [here](https://github.com/coreos/coreos-assembler/compare/f33f7a06ca6bd5ee427adba2cc3891498c6b0076..ac4b48e48990097ec24426a1c4035a3fe0d1f004)

phoracek commented on mantle/kola/tests/misc/network.go:
```diff
@@ -334,6 +504,29 @@ func getBondIfaceMAC(c cluster.TestCluster, machineIdx int, interfaceName string
 }
 
+func checkExpectedIP(c cluster.TestCluster, m platform.Machine, interfaceName, expectedIP string) {
+	InterfaceIPAddress, err := getInterfaceIP(c, m, interfaceName)
```
```suggestion
	interfaceIPAddress, err := getInterfaceIP(c, m, interfaceName)
```

phoracek commented on mantle/kola/tests/misc/network.go:
```diff
@@ -334,6 +504,29 @@ func getBondIfaceMAC(c cluster.TestCluster, machineIdx int, interfaceName string
+		c.Fatalf("failed to fetch bond IP Address: %v", err)
+	}
+
+	if InterfaceIPAddress != expectedIP {
```
```suggestion
	if interfaceIPAddress != expectedIP {
```

RamLavi replied:
ahh good catch, Done

phoracek reviewed (commented):
Just in a nit, looks good overall

RamLavi commented:
@phoracek fix [here](https://github.com/coreos/coreos-assembler/compare/ac4b48e48990097ec24426a1c4035a3fe0d1f004..8c289d76c84b3bf6922412314f7507ca6eaa9d05)

jlebon reviewed (approved):
LGTM.

Can you confirm you've successfully tested this latest version on RHCOS?

RamLavi commented:
> LGTM.
> 
> Can you confirm you've successfully tested this latest version on RHCOS?

Yes, I've tested this locally, both tests pass
```
=== RUN   rhcos.network.multiple-nics
=== RUN   rhcos.network.bond-with-dhcp
--- PASS: rhcos.network.multiple-nics (77.15s)
--- PASS: rhcos.network.bond-with-dhcp (76.69s)
PASS, output in tmp/kola/qemu-unpriv-2021-06-28-1146-2287764
```

Moreover, In order to keep the [rhcos-slb](https://github.com/RHsyseng/rhcos-slb/) repo always in check, I've added an [issue](https://github.com/RHsyseng/rhcos-slb/issues/23) & [PR](https://github.com/RHsyseng/rhcos-slb/pull/24) to run these coreos tests locally on future changes that might be done on rhcos-slb
